packages/govuk-frontend/src/govuk/components/password-input/password-input.mjs

```diff
@@ -0,0 +1,265 @@
+ *
+ * @preserve
+ */
+export class PasswordInput extends GOVUKFrontendComponent {
+  /** @private */
+  $module
+
+  /**
+   * @private
+   * @type {PasswordInputConfig}
+   */
+  config
+
+  /** @private */
+  i18n
+
+  /**
+   * @private
+   * @type {HTMLInputElement}
+   */
+  $input
+
+  /**
+   * @private
+   * @type {HTMLButtonElement}
+   */
+  $showHideButton
+
+  /** @private */
+  $statusText
+
+  /**
+   * @param {Element} $module - HTML element to use for password input
+   * @param {PasswordInputConfig} [config] - Password input config
+   */
+  constructor($module, config = {}) {
+    super()
+
+    if (!($module instanceof HTMLElement)) {
+      throw new ElementError({
+        componentName: 'Password input',
+        element: $module,
+        identifier: 'Root element (`$module`)'
+      })
+    }
+
+    const $input = $module.querySelector('.govuk-js-password-input-input')
+    if (!($input instanceof HTMLInputElement)) {
+      throw new ElementError({
+        componentName: 'Password input',
+        element: $input,
+        expectedType: 'HTMLInputElement',
+        identifier: 'Form field (`.govuk-js-password-input-input`)'
+      })
+    }
+
+    const $showHideButton = $module.querySelector(
+      '.govuk-js-password-input-toggle'
+    )
+    if (!($showHideButton instanceof HTMLButtonElement)) {
+      throw new ElementError({
+        componentName: 'Password input',
+        element: $showHideButton,
+        expectedType: 'HTMLButtonElement',
+        identifier: 'Button (`.govuk-js-password-input-toggle`)'
+      })
+    }
+
+    this.$module = $module
+    this.$input = $input
+    this.$showHideButton = $showHideButton
+
+    this.config = mergeConfigs(
+      PasswordInput.defaults,
+      config,
+      normaliseDataset($module.dataset)
+    )
+
+    // Check for valid config
+    const errors = validateConfig(PasswordInput.schema, this.config)
+    if (errors[0]) {
+      throw new ConfigError(`Password input: ${errors[0]}`)
+    }
+
+    this.i18n = new I18n(extractConfigByNamespace(this.config, 'i18n'), {
+      // Read the fallback if necessary rather than have it set in the defaults
+      locale: closestAttributeValue($module, 'lang')
+    })
+
+    // Show the toggle button element
+    this.$showHideButton.removeAttribute('hidden')
+
+    // Create and append the status text for screen readers.
+    // This is injected between the input and button so that users get a sensible reading order if
+    // moving through the page content linearly:
+    // [password input] -> [your password is visible/hidden] -> [show/hide password]
+    this.$statusText = document.createElement('span')
+    this.$statusText.className = 'govuk-visually-hidden'
+    this.$statusText.innerText = this.i18n.t('passwordHiddenAnnouncement')
+    this.$statusText.setAttribute('aria-live', 'polite')
+    this.$input.parentNode?.insertBefore(
+      this.$statusText,
+      this.$input.nextSibling
+    )
```

Comment:
Wondering if we should align with Character count here?

1. Naming convention `$screenReaderXXXMessage`
2. Updates via `updateScreenReaderXXXMessage()`
3. Element class `__sr-status` used in test selectors
4. Element as `<div>` rather than `<span>`
5. Injected via `insertAdjacentElement()`

https://github.com/alphagov/govuk-frontend/blob/56fc909f395caede98440e061488bddc940fef81/packages/govuk-frontend/src/govuk/components/character-count/character-count.mjs#L159-L169

Reply:
Or whether we should abstract both away into a utility, base class or Nunjucks macro?

Reply:
Probably sensible to align variable names and classes closer to how the character count has it, agreed. 

I don't think 2 is needed in this case, character count only does has a separate method because it additionally has to account for the threshold functionality and to differentiate it from the separate visible counter. 

Two mostly-similar cases doesn't constitute a trend in my mind, I think abstracting them might be a premature optimisation.